### PR TITLE
Valid and Same padding option in convolution layers.

### PR DESCRIPTION
**Pull request by walragatver**
Hi,
I have modified `convolutional` layer. But I have a doubt. Should we take all four padding dimension's as an argument in `convolution.hpp`. Currently we only take two dimension as argument in `constructor`. I have kept it as it is for now. I think taking all four padding dimension's would be good but it may break backward compatibility. Let me know what you think.

### CODE REVIEW

File: src/mlpack/methods/ann/layer/convolution.hpp [hunk 1]
```diff
@@ -74,7 +75,8 @@ class Convolution
               const size_t padW = 0,
               const size_t padH = 0,
               const size_t inputWidth = 0,
-              const size_t inputHeight = 0);
+              const size_t inputHeight = 0,
+              const std::string paddingType = "None");
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/convolution.hpp]: I'm wondering if implementing another template argument (policy) would be more consistent with the existing mlpack code, let me know what you think. I don't have a preference here.

File: src/mlpack/methods/ann/layer/convolution_impl.hpp [hunk 2]
```diff
@@ -118,10 +133,10 @@ void Convolution<
   inputTemp = arma::cube(const_cast<arma::Mat<eT>&&>(input).memptr(),
       inputWidth, inputHeight, inSize * batchSize, false, false);
 
-  if (padW != 0 || padH != 0)
+  if (padWLeft != 0 || padWRight != 0 || padHTop != 0 || padHBottom != 0)
   {
-    inputPaddedTemp.set_size(inputTemp.n_rows + padW * 2,
-        inputTemp.n_cols + padH * 2, inputTemp.n_slices);
+    inputPaddedTemp.set_size(inputTemp.n_rows + padWLeft + padWRight,
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/convolution_impl.hpp]: Would be great if we could add a test as well.

**Comment by zoq** [comment]
We could implement another constructor that takes the other dimensions as well, that way we could be backward compatible.

File: src/mlpack/methods/ann/layer/convolution.hpp (hunk 1 above)
Comment (walragatver) [reply src/mlpack/methods/ann/layer/convolution.hpp]: It looks like we cannot use `string literals` as template arguments. Should I create new classes for policies like we did for `GAN` to have `StandardGan` and `DCGAN` as class?

Comment (walragatver) [reply src/mlpack/methods/ann/layer/convolution.hpp]: Currently I am keeping it same. It looks like I will need to create classes for it as a template argument. Let me know if it's incorrect. I will change it.

File: src/mlpack/methods/ann/layer/convolution_impl.hpp [hunk 3]
```diff
@@ -92,6 +92,62 @@ Convolution<
...
+    padWRight(std::get<1>(padW)),
+    padHBottom(std::get<1>(padH)),
+    padHTop(std::get<0>(padH)),
+    inputWidth(inputWidth),
+    inputHeight(inputHeight),
+    outputWidth(0),
+    outputHeight(0)
+{
+  weights.set_size((outSize * inSize * kW * kH) + outSize, 1);
+  if (paddingType == "Valid")
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/convolution_impl.hpp]: Perhaps we should transform ``paddingType`` to lower case and compare with ``"valid"`` what do you think?

File: src/mlpack/methods/ann/layer/convolution.hpp (hunk 1 above)
Comment (zoq) [reply src/mlpack/methods/ann/layer/convolution.hpp]: I think a string is fine, even closer to tf/keras.

File: src/mlpack/methods/ann/layer/convolution_impl.hpp (hunk 3 above)
Comment (walragatver) [reply src/mlpack/methods/ann/layer/convolution_impl.hpp]: Yes done.

File: src/mlpack/methods/ann/layer/convolution_impl.hpp (hunk 2 above)
Comment (walragatver) [reply src/mlpack/methods/ann/layer/convolution_impl.hpp]: Actually I am doing this for `Inception Layer`, so I think we will know whether it is wrong or not. Logically It seems correct. Let me know what you think.

Comment (zoq) [reply src/mlpack/methods/ann/layer/convolution_impl.hpp]: Sorry for the slow response, ideally we can write an independent test for the adjusted conv operation, similar to https://github.com/mlpack/mlpack/blob/444559d71e1d29ef518c836aa571b40850a8755c/src/mlpack/tests/convolution_test.cpp#L124.

**Comment by walragatver** [comment]
Hi @zoq, I will add test soon. Sorry for the delay.

**Comment by walragatver** [comment]
Hi @zoq,
I have added the test. I am thinking to work on `transposed convolution` in a separate PR. I have added tests in `ann_layer_test.cpp` file. Let me know if it's not the correct position.

**Comment by zoq** [comment]
Okay, will take a look, thanks.

**Comment by walragatver** [comment]
Also, 
Can I add this to the next milestone as @sreenikSS, pointed out its a necessity. I think only a review is required for this as well.

File: src/mlpack/tests/ann_layer_test.cpp [hunk 4]
```diff
@@ -2729,4 +2770,42 @@ BOOST_AUTO_TEST_CASE(LayerNormSerializationTest)
...
+  arma::mat output, input, delta;
+
+  // Check valid padding option.
+  Convolution<> module1(1, 1, 3, 3, 1, 1, {1, 1}, {1, 1}, 7, 7, "valid");
+
+  // Test the Forward function.
+  input = arma::linspace<arma::colvec>(0, 48, 49);
+  module1.Parameters() = arma::mat(9 + 1, 1, arma::fill::zeros);
+  module1.Reset();
+  module1.Forward(std::move(input), std::move(output));
```
Comment (zoq) [inline src/mlpack/tests/ann_layer_test.cpp]: Would be a good idea to check the output as well, but if you are going to test this is combination with the  Inception layer this is fine as well.

Comment (walragatver) [reply src/mlpack/tests/ann_layer_test.cpp]: Yes, you are correct. I am now checking the `sum of all values of output as zero`. Let me know if anything else is needed to be added.

**Comment by walragatver** [comment]
Hi @zoq,
It looks like the build is failing because of this. The build was passing in my machine as I was using C++14. What should we do now? Can we upgrade our `C++` version while building? I am not sure what I am saying let me know what you think.
https://stackoverflow.com/questions/32084706/why-initialization-doesnt-work-for-tuple

**Comment by rcurtin** [comment]
@walragatver it's a little bit uglier, but you could do `std::tuple<size_t, size_t>(1, 1)` and that should work. :+1:  Given how old some of the compilers are for still-in-use distros like RHEL6, I don't think that we should change to require C++14 quite yet.
